### PR TITLE
Fixed #337, #344 & improved logging error situations

### LINKED ISSUE #337: Custom headers in knot.x HTTP calls
As of right now it's not possible to determine whether request comes from knot.x or some other HTTP client (w/o extra work on different layers), as by default it transparently passes on all the headers from original request.

A few use cases:
* HTTP routing for repository requests (request source is taken into consideration)
* integration with external services (i.e. some API may require extra HTTP header for the sake of authentication and/or tracking purposes)

In general it'd be great to have an ability to add headers to any HTTP call that knot.x does. I'm also thinking about generic header attached to every request that contains metadata about the source (something similar to `User-Agent`).

### PR DESCRIPTION
**Pull request by marcinczeczko**
- better logging for Server routing, http service adapter, http respository connector
- added custom header response/request - to help easier create an infrastructure
- small polisher of other aspects :)

### CODE REVIEW

File: knotx-adapter/knotx-adapter-common/src/main/java/io/knotx/adapter/common/http/HttpAdapterConfiguration.java [hunk 1]
```diff
@@ -15,20 +15,30 @@
...
+
   private String address;
 
   private List<ServiceMetadata> services;
 
   private JsonObject clientOptions;
 
+  public HttpAdapterConfiguration() {
+    customRequestHeader = new JsonObject();
+    services = Lists.newArrayList();
```
Comment (Skejven) [inline knotx-adapter/knotx-adapter-common/src/main/java/io/knotx/adapter/common/http/HttpAdapterConfiguration.java]: Can we set `Collections.emptyList()` here?

File: knotx-adapter/knotx-adapter-common/src/main/java/io/knotx/adapter/common/http/HttpClientFacade.java [hunk 2]
```diff
@@ -49,23 +51,44 @@
...
-        .flatMap(serviceRequest -> callService(serviceRequest, method))
+        .flatMap(
+            serviceRequest -> callService(serviceRequest, method)
+                .doOnSuccess(resp -> logResponse(serviceRequest, resp)))
         .flatMap(this::wrapResponse);
   }
 
+  private void logResponse(Pair<ClientRequest, ServiceMetadata> request,
+      HttpResponse<Buffer> resp) {
+    if (resp.statusCode() != HttpResponseStatus.OK.code()) {
```
Comment (Skejven) [inline knotx-adapter/knotx-adapter-common/src/main/java/io/knotx/adapter/common/http/HttpClientFacade.java]: Not sure if we should `WARN` about other status codes than `200`.
Statuses other than `200` are sometimes used to communicate status of REST operation (e.g. like `201`, `203` or `204`).
Maybe this should be `DEBUG` level here or `WARN` only `4XX` and `5XX` responses and `TRACE` others?

File: knotx-adapter/knotx-adapter-common/src/main/java/io/knotx/adapter/common/http/HttpClientFacade.java [hunk 3]
```diff
@@ -77,17 +100,12 @@ protected void validateContract(AdapterRequest message) {
...
-   * <br>
-   * The responsibility of the method is to build ClientRequest based on the original Http
-   * Request<br>
-   * - It must set path property of the request based on the params<br>
-   * - It might set headers of the request if needed.<br>
-   * <br>
-   * In case of headers created modified in this method, ensure that your service configuration
-   * allows passing those headers to the target service. See 'allowedRequestHeaders' section
-   * of the configuration <br>
+   * Method responsible for building request to the service. <br> <br> The responsibility of the
```
Comment (Skejven) [inline knotx-adapter/knotx-adapter-common/src/main/java/io/knotx/adapter/common/http/HttpClientFacade.java]: Can you please bring back formatting ?

File: knotx-adapter/knotx-adapter-common/src/main/java/io/knotx/adapter/common/http/HttpClientFacade.java (hunk 2 above)
Comment (tomaszmichalak) [reply knotx-adapter/knotx-adapter-common/src/main/java/io/knotx/adapter/common/http/HttpClientFacade.java]: Agree.

File: knotx-knot/knotx-knot-service/src/main/java/io/knotx/knot/service/service/ServiceEngine.java [hunk 4]
```diff
@@ -83,7 +85,9 @@ private JsonObject buildResultObject(AdapterResponse adapterResponse) {
     } else if (rawData.charAt(0) == '{') {
       object.put(RESULT_NAMESPACE_KEY, new JsonObject(rawData));
     } else {
-      throw new DecodeException("Result is neither Json Array nor Json Object");
+      LOGGER.error("Result of [{} {}] neither Json Array nor Json Object: [{}]",
```
Comment (tomaszmichalak) [inline knotx-knot/knotx-knot-service/src/main/java/io/knotx/knot/service/service/ServiceEngine.java]: Should we hide this exception here? It changes a service error scenario behaviour.

File: knotx-repository-connector/knotx-repository-connector-http/src/main/java/io/knotx/repository/impl/RepositoryConnectorProxyImpl.java [hunk 5]
```diff
@@ -143,22 +146,44 @@ private String encodeParamValue(String value) {
...
 
-  private ClientResponse toSuccessResponse(Buffer buffer,
-      final HttpClientResponse httpClientResponse) {
-    return new ClientResponse()
-        .setStatusCode(httpClientResponse.statusCode())
-        .setHeaders(httpClientResponse.headers())
-        .setBody((io.vertx.core.buffer.Buffer) buffer.getDelegate());
+  private ClientResponse toResponse(Buffer buffer, final HttpClientResponse httpResponse) {
+    if (httpResponse.statusCode() >= 300 && httpResponse.statusCode() < 400) { //redirect responses
+      LOGGER.warn("Repository 3xx response: {}, Headers[{}]", httpResponse.statusCode(),
```
Comment (tomaszmichalak) [inline knotx-repository-connector/knotx-repository-connector-http/src/main/java/io/knotx/repository/impl/RepositoryConnectorProxyImpl.java]: info?

File: documentation/src/main/wiki/HttpRepositoryConnector.md [hunk 6]
```diff
@@ -27,24 +27,31 @@ Default configuration shipped with the verticle as `io.knotx.HttpRepositoryConne
         "domain": "localhost",
         "port": 3001
       },
+      "customRequestHeader": {
```
[the remark this answers is not in the document]
Comment (Skejven) [reply documentation/src/main/wiki/HttpRepositoryConnector.md]: Actually, we had discussion about it with @marcinczeczko . The reason here why it is just a single item, is that this mechanism should be just a marker about the origin of a request. All necessary logic connected with headers manipulation should be done at the different layer (e.g. Apache/Ngnix).

File: knotx-knot/knotx-knot-service/src/main/java/io/knotx/knot/service/service/ServiceEngine.java (hunk 4 above)
Comment (marcinczeczko) [reply knotx-knot/knotx-knot-service/src/main/java/io/knotx/knot/service/service/ServiceEngine.java]: We're not hiding anything here ? Before it was the exception generated here that broke the whole processing. But in fact instead it should log the error that data is not parsable and return empty json.

File: documentation/src/main/wiki/HttpRepositoryConnector.md [hunk 7]
```diff
@@ -27,24 +27,31 @@ Default configuration shipped with the verticle as `io.knotx.HttpRepositoryConne
...
+        "name": "Server-User-Agent",
+        "value": "Knot.x"
+      },
       "allowedRequestHeaders": [
```
[the remark this answers is not in the document]
Comment (marcinczeczko) [reply documentation/src/main/wiki/HttpRepositoryConnector.md]: Now we shouldn't. If it's defined it's always to be send.

File: knotx-adapter/knotx-adapter-common/src/main/java/io/knotx/adapter/common/http/HttpClientFacade.java [hunk 8]
```diff
@@ -49,19 +50,40 @@
...
-        .flatMap(serviceRequest -> callService(serviceRequest, method))
+        .flatMap(
+            serviceRequest -> callService(serviceRequest, method)
+                .doOnSuccess(resp -> logResponse(serviceRequest, resp)))
         .flatMap(this::wrapResponse);
   }
 
+  private void logResponse(Pair<ClientRequest, ServiceMetadata> request,
+                           HttpResponse<Buffer> resp) {
+    if (resp.statusCode() >= 400 && resp.statusCode() < 600) {
```
Comment (Skejven) [inline knotx-adapter/knotx-adapter-common/src/main/java/io/knotx/adapter/common/http/HttpClientFacade.java]: Do we want to `trace` other status codes?

Comment (marcinczeczko) [reply knotx-adapter/knotx-adapter-common/src/main/java/io/knotx/adapter/common/http/HttpClientFacade.java]: Good point. I'll check and will add trace for it and then do the merge

**Pull request merged**